privacyidea/lib/queue/collector.py

```diff
@@ -0,0 +1,92 @@
+# version 3 of the License, or any later version.
+#
+# This code is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU AFFERO GENERAL PUBLIC LICENSE for more details.
+#
+# You should have received a copy of the GNU Affero General Public
+# License along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+#
+
+import logging
+
+log = logging.getLogger(__name__)
+
+
+JOB_QUEUE_CLASS = "PI_JOB_QUEUE_CLASS"
+JOB_QUEUE_OPTION_PREFIX = "PI_JOB_QUEUE_"
+
+
+class JobCollector(object):
+    """
+    For most third-party job queue modules, the jobs are discovered by tracking all
+    functions decorated with a ``@job`` decorator. However, in order
+    to invoke the decorator, one usually needs to provide the queue
+    configuration (e.g. the redis server) already.
+    In privacyIDEA, we cannot do that, because the app config is not known
+    yet -- it will be known when ``create_app`` is called!
+    Thus, we cannot directly use the @job decorator, but need a job collector
+    that collects jobs in privacyIDEA code and registers them with the job
+    queue module when ``create_app`` has been called.
+    """
+    def __init__(self, queue_classes, default_queue_class_name):
+        """
+        :param queue_classes: A dictionary mapping class names to BaseQueue sclasses
+        :param default_queue_class_name: Class name to use if the config specifies no queue class
+        """
+        self._jobs = {}
+        self._queue_classes = queue_classes
+        self._default_queue_class_name = default_queue_class_name
+
+    @property
+    def jobs(self):
+        return self._jobs
+
+    def add_job(self, name, func, args, kwargs):
+        """
+        Register a job with the collector.
+        :param name: unique name of the job
+        :param func: function of the job
+        :param args: arguments passed to the job queue's ``add_job`` method
+        :param kwargs: keyword arguments passed to the job queue's ``add_job`` method
+        """
+        if name in self._jobs:
+            raise RuntimeError("Duplicate jobs: {!r}".format(name))
+        self._jobs[name] = (func, args, kwargs)
+
+    def register_app(self, app):
+        """
+        Create an instance of a ``BaseQueue`` subclass according to the app config's
+        ``PI_JOB_QUEUE_CLASS`` option and store it in the ``job_queue`` config.
+        Register all collected jobs with this application.
+        This instance is shared between threads!
+        This function should only be called once per process.
+        :param app: privacyIDEA app
+        """
+        if "job_queue" in app.config:
+            raise RuntimeError("App already has a job queue: {!r}".format(app.config["job_queue"]))
+        queue_class_name = app.config.get(JOB_QUEUE_CLASS, self._default_queue_class_name)
```

Comment:
I agree, having shortnames at some places and class names at other places is quite inconsistent and not very nice :-/ I've also opened #1337 for a similar problem.

Actually I think in the case of queue classes, it would be quite nice to have the possibility to load external queue classes. So maybe we should change the ``PI_JOB_QUEUE_CLASS`` option to accept a class name instead of a shortname?

Reply:
Thanks

Reply:
I changed the ``PI_JOB_QUEUE_CLASS`` option to accept a module + class name instead of a shortname. e.g. to use the huey-based job queue, one needs to set
```
PI_JOB_QUEUE_CLASS = 'privacyidea.lib.queues.huey_queue.HueyQueue'
```